plugins/pixart-rf/fu-pxi-ble-device.c

```diff
@@ -185,49 +190,108 @@ fu_pxi_ble_device_get_feature(FuPxiBleDevice *self, guint8 *buf, guint bufsz, GE
 }
 
 static gboolean
-fu_pxi_ble_device_search_hid_usage_page(guint8 *report_descriptor,
+fu_pxi_ble_device_search_hid_feature_report_id(guint8 *report_descriptor,
 					gint size,
-					guint8 *usage_page,
-					guint8 usage_page_sz)
+					guint16 usage_page,
+					guint8 *report_id,
+					GError **error)
 {
-	gint pos = 0;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(FuHidDescriptor) descriptor = fu_hid_descriptor_new();
+	g_autoptr(FuHidReport) report = NULL;
+	g_autoptr(FuFirmware) item_id = NULL;
 
-	fu_dump_raw(G_LOG_DOMAIN, "target usage_page", usage_page, usage_page_sz);
+	/* Init the Descriptor */
+	fw = g_bytes_new(report_descriptor, size);
 
-	while (pos < size) {
-		/* HID info define by HID specification */
-		guint8 item = report_descriptor[pos];
-		guint8 report_size = item & 0x03;
-		guint8 report_tag = item & 0xF0;
-		guint8 usage_page_tmp[4] = {0x00};
+	/* Preprocess the Descriptor */
+	if (!fu_firmware_parse(descriptor, fw, FWUPD_INSTALL_FLAG_NONE, error))
+		return FALSE;
 
-		report_size = (report_size == 3) ? 4 : report_size;
+	/* check ota retransmit feature report usage page exist or not */
+	report = fu_hid_descriptor_find_report(descriptor,
+					       error,
+					       "usage-page",
+					       usage_page,
+					       "usage",
+					       0x01,
+					       "feature",
+					       0x02,
+					       NULL);
+	if (report == NULL)
+		return FALSE;
 
-		if (report_tag != 0) {
-			pos += report_size + 1;
-			continue;
-		}
+	/* find report-id */
+	item_id = fu_firmware_get_image_by_id(FU_FIRMWARE(report), "report-id", error);
 
-		memmove(usage_page_tmp, &report_descriptor[pos + 1], report_size);
-		if (memcmp(usage_page, usage_page_tmp, usage_page_sz) == 0) {
-			g_debug("hit item: %x  ", item);
-			fu_dump_raw(G_LOG_DOMAIN, "usage_page", usage_page, report_size);
-			g_debug("hit pos %d", pos);
-			return TRUE; /* finished processing */
-		}
-		pos += report_size + 1;
-	}
+	if (item_id == NULL)
+		return FALSE;
+
+	/* find report-id value */
+	*report_id = fu_hid_report_item_get_value(FU_HID_REPORT_ITEM(item_id));
+
+	g_debug("usage-page: %x report_id:%d",usage_page, *report_id);
 
 	return FALSE; /* finished processing */
```

Comment:
nah, if you return FALSE you need to set GError. I think you want to return TRUE here.